medmodels/matching/matching.py

```diff
@@ -0,0 +1,40 @@
+"""Matching class template for patient matching algorithms"""
+
+from typing import Set
+
+from medmodels.medrecord.medrecord import MedRecord
+from medmodels.medrecord.types import NodeIndex
+
+
+class Matching:
+    """Base class for Patient Matching algorithms"""
+
+    def __init__(
+        self,
+        medrecord: MedRecord,
+        treated_group: Set[NodeIndex],
+        control_group: Set[NodeIndex],
+    ) -> None:
+        """
+        Initializes the Matching object with the MedRecord object and the treated and
+        control groups.
+
+        Args:
+            medrecord (MedRecord): An instance of the MedRecord class containing patient
+                medical data.
+            treated_group (Set[NodeIndex]): A set containing the IDs of patients in the
+                treated group.
+            control_group (Set[NodeIndex]): A set containing the IDs of patients in the
+                control group.
+
+        Returns:
+            None
+        """
+        self.medrecord = medrecord
+        assert isinstance(medrecord, MedRecord), "medrecord must be a MedRecord object"
```

Comment:
Asserting that the medrecord really is a medrecord is unnecessary and already covered by the type system. We do not check for user error only when the user can reach an unwanted state while following our function typing.